src/Microsoft.DotNet.ImageBuilder/src/Commands/GenerateEolAnnotationDataCommand.cs

```diff
@@ -0,0 +1,235 @@
+                registryIndex++;
+            }
+            else
+            {
+                supportedIndex++;
+                registryIndex++;
+            }
+        }
+
+        // If there are remaining items, add them to the result
+        while (registryIndex < sortedRegistryDigests.Count)
+        {
+            (string digest, string tag) = sortedRegistryDigests[registryIndex];
+            result.Add(new EolDigestData(digest) { Tag = tag });
+            registryIndex++;
+        }
+
+        return result;
+    }
+
+    private IEnumerable<string> GetSupportedDigests() =>
+        _newImageArtifactDetails.Repos
+            .SelectMany(repo => repo.Images)
+            .SelectMany(image => GetImageDigests(image))
+            .Select(digest => digest.Digest);
+    
+    private static IEnumerable<(string Digest, string? Tag)> GetImageDigests(ImageData image)
+    {
+        if (image.Manifest is not null)
+        {
+            yield return (image.Manifest.Digest, GetLongestTag(image.Manifest.SharedTags));
+        }
+
+        foreach (PlatformData platform in image.Platforms)
+        {
+            yield return (platform.Digest, GetLongestTag(platform.SimpleTags));
+        }
+    }
+
+    private static string? GetLongestTag(IEnumerable<string> tags) =>
+        tags.OrderByDescending(tag => tag.Length).FirstOrDefault();
+
+    private async Task<IEnumerable<(string Digest, string? Tag)>> GetAllDigestsFromRegistry(IEnumerable<string> repoNames)
+    {
+        IContainerRegistryClient acrClient = _acrClientFactory.Create(Options.RegistryName, _tokenCredentialProvider.GetCredential());
+        IAsyncEnumerable<string> repositoryNames = acrClient.GetRepositoryNamesAsync();
+
+        ConcurrentBag<(string Digest, string? Tag)> digests = [];
+        await foreach (string repositoryName in repositoryNames.Where(name => repoNames.Contains(name)))
+        {
+            ContainerRepository repo = acrClient.GetRepository(repositoryName);
+            IAsyncEnumerable<ArtifactManifestProperties> manifests = repo.GetAllManifestPropertiesAsync();
+            await foreach (ArtifactManifestProperties manifestProps in manifests)
+            {
+                string imageName = DockerHelper.GetImageName(Options.RegistryName, repositoryName, digest: manifestProps.Digest);
+                digests.Add((imageName, GetLongestTag(manifestProps.Tags)));
+            }
+        }
+
+        return digests;
+    }
+
+    private IEnumerable<EolDigestData> GetProductEolDigests(ImageData image)
+    {
+        if (image.ProductVersion != null)
+        {
+            string dotnetVersion = Version.Parse(image.ProductVersion).ToString(2);
+            if (_productEolDates != null && _productEolDates.TryGetValue(dotnetVersion, out DateOnly? date))
+            {
+                return GetImageDigests(image).Select(val => new EolDigestData(val.Digest) { Tag = val.Tag, EolDate = date });
+            }
+        }
+
+        return [];
```

Comment:
Consider inverting these conditions and returning early?

```suggestion
        if (image.ProductVersion == null)
        {
            return [];
        }

        string dotnetVersion = Version.Parse(image.ProductVersion).ToString(2);

        if (_productEolDates == null || !_productEolDates.TryGetValue(dotnetVersion, out DateOnly? date))
        {
            return [];
        }

        return GetImageDigests(image).Select(val => new EolDigestData(val.Digest) { Tag = val.Tag, EolDate = date });
```